client/lib/cart-values/index.js

```diff
@@ -16,6 +16,53 @@ import cartItems from './cart-items';
 import productsValues from 'lib/products-values';
 import { requestGeoLocation } from 'state/data-getters';
 
+/**
+ * Preprocesses cart for server.
+ *
+ * @param {Object} cart Cart object.
+ * @returns {Object} A new cart object.
+ */
+function preprocessCartForServer( {
+	coupon,
+	is_coupon_applied,
+	is_coupon_removed,
+	currency,
+	temporary,
+	extra,
+	products,
+} ) {
+	const needsUrlCoupon = ! (
+		coupon ||
+		is_coupon_applied ||
+		is_coupon_removed ||
+		typeof document === 'undefined'
+	);
+	const urlCoupon = needsUrlCoupon ? url.parse( document.URL, true ).query.coupon : '';
+
+	return Object.assign(
+		{
+			coupon,
+			is_coupon_applied,
+			is_coupon_removed,
+			currency,
+			temporary,
+			extra,
+			products: products.map( ( { product_id, meta, free_trial, volume, _extra } ) => ( {
```

Comment:
Noting here that this `_extra` bit has broken G Suite purchases. Fix in https://github.com/Automattic/wp-calypso/pull/25713/. I'd like to also understand how this sneaked in through code review - such a big change to shopping cart should have been thoroughly tested, with all the possible purchases. Would be great to ping relevant teams as well (Cobalt for domain purchases and Neutron for purchases in general).
The testing instructions don't even mention verifying that the shopping cart is working as it should after these changes - just that the coupon has been applied. Please be more careful when modifying shopping cart logic in the future - it's very brittle (I know, I've been burned by it quite a few times) and extra eyes are always welcome (and I'm always happy to code review!).

Reply:
Points taken and I will be more careful in the future.
This was my fault. I will write up a post-mortem about this now.